### PR TITLE
Replace `privileged_tests` build tag with `PRIVILEGED_TESTS` environment variable

### PR DESCRIPTION
**Pull request by ti-mo**
This is going to ping all SIGs, sorry about that. This PR replaces the `privileged_tests` build tag with test helpers that skip the test or check unless the the `PRIVILEGED_TESTS` environment variable is set.

Using build tags for splitting up a test suite and code in general has a few drawbacks:

- Multiple `go build`s and `go test`s need to be run to ensure all code compiles. This compounds as the matrix of target GOOS and GOARCH values grows, which it has over the past few months/years.
- gopls needs to be configured with -tags=privileged_tests, but now you no longer get code completion in non-privileged tests, which is the majority of files in the codebase.
- Same goes for linters and formatters that are build-tag aware, like golangci-lint and goimports. One needs to pick for which build tags they want to receive warnings and errors.

Why not test flags instead? Here's [an opinion piece](https://peter.bourgon.org/blog/2021/04/02/dont-use-build-tags-for-integration-tests.html) on why environment variables are preferred. The article raises some valid points, but the real case against using test flags is that they need to be declared in every single directory containing a .go file. If there's a package without a flag defined, this is the result:

```
~ go test ./... -run xxxx -privileged
...
flag provided but not defined: -privileged
FAIL	github.com/cilium/cilium/bugtool/cmd	0.006s
```

A three-line test helper is all we really need to accomplish this with an environment variable. Moreover, env vars are easy to add to a build/test context from a Makefile, and they will automatically apply to any and all child processes. Marking privileged tests using a helper gives us the flexibility of changing the underlying mechanism of the helper. This PR proposes and an env var, but we could just as well use another trigger down the line like checking the current UID or checking caps using `capsh`, though that can cause tests to run unintentionally.

```release-note
Replace `privileged_tests` build tag with `PRIVILEGED_TESTS` environment variable
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -125,17 +125,13 @@ build-container: check-sources ## Builds components required for cilium-agent co
...
-tests-privileged:
+tests-privileged: ## Run integration tests for Cilium that require elevated privileges.
 	$(MAKE) init-coverage
-	for pkg in $(patsubst %,github.com/cilium/cilium/%,$(PRIV_TEST_PKGS)); do \
+	for pkg in $(patsubst %,github.com/cilium/cilium/%,$(TESTPKGS)); do \
 		>&2 $(ECHO_TEST) $$pkg; \
-		PATH=$(PATH):$(ROOT_DIR)/bpf $(GO_TEST) $(TEST_LDFLAGS) $$pkg $(GOTEST_UNIT_BASE) $(GOTEST_COVER_OPTS) -coverpkg $$pkg \
+		PATH=$(PATH):$(ROOT_DIR)/bpf \
+		PRIVILEGED_TESTS=true \
+		$(GO_TEST) $(TEST_LDFLAGS) $$pkg $(GOTEST_UNIT_BASE) $(GOTEST_COVER_OPTS) -coverpkg $$pkg \
```
Comment (joestringer) [inline Makefile]: Does this mean that the `tests-privileged` now runs _all_ tests, not just the privileged ones?

Currently we run the unprivileged tests in Travis and run the privileged tests in the ginkgo/jenkins environments, so if this target now includes unprivileged tests then it will likely extend the test runtime by 5 minutes or so, for results we already get from another mechanism. Do you think there's a way we can avoid this with the new mechanism?

**Review by joestringer — COMMENTED** [review]
There's some nice properties provided by this proposal :+1: 

My main concern is around extending the testsuite runtime by running more tests when we don't need to. Comment below.

File: Makefile (hunk 1 above)
Comment (joestringer) [reply Makefile]: Couple of possible areas to explore:
- Ensure there is a way to skip unprivileged tests, so that jenkins test doesn't need to run those additional tests that are already run elsewhere
- Move privileged tests out of the jenkins CI into a github action, so we don't extend the jenkins test job runtimes but instead the privileged tests run in parallel with other jobs

Comment (ti-mo) [reply Makefile]: @joestringer Thanks for taking a look! I had all the points you raised in mind while working on this.

First,

> Does this mean that the `tests-privileged` now runs all tests, not just the privileged ones?

Yes! And that's fine, because unit ('unprivileged') tests are actually very fast to run, as they should be. The most expensive part of running our test suite is building test binaries. Being able to run all unit, privileged and integration tests in one go is one of the end goals of these changes.

With build cache (4.38s):
```
~ time go test ./...
...
go test ./...  29.60s user 4.38s system 763% cpu 4.450 total
```

Without build cache:
```
~ go clean -testcache ./...
~ time go test ./...
go test ./...  245.62s user 25.07s system 835% cpu 32.383 total
```

This is on a Ryzen 3700X (16 threads) desktop CPU with fast memory, with all source files in OS cache. Even on this machine, build time accounts for 87% of total test suite runtime. I would expect this number to increase even further on cloud VMs with noisy neighbours and slower i/o.

IMO the only proper way of skipping unprivileged tests would be by consistenly naming test functions, e.g. `func Test(Un)PrivilegedFoo()` so they can be filtered out by `go test -run=TestPrivileged`. This is not a realistic goal given the amount of testing frameworks in use throughout the codebase that all have their own quirky way of doing things.

In conclusion: always running unit tests is fine and barely amounts to a rounding error on a total privileged test runtime of 2.5+ hours. It doesn't weigh up against renaming all test functions in existence, nor to adding test helpers to all of them.

---

Second, regarding Jenkins/Travis/GHA, ...

For me, the next step would be cutting out ginkgo from running privileged tests on Jenkins. As far as I can see, it's currently only used for invoking vagrant. The code is inscrutable at best, and heavily interferes in runtime test output, especially when failures occur.

Note that we need nested virt for privileged tests, so we'd need self-hosted GHA runners first.

We'll get there, this is just a first attempt at moving the needle. :slightly_smiling_face:

File: .golangci.yaml [hunk 2]
```diff
@@ -13,8 +13,8 @@ run:
   tests: true
 
   # list of build tags, all linters use it. Default is empty list.
-  build-tags:
```
Comment (nebril) [inline .golangci.yaml:16]: Should we just remove it?

**Review by nebril — APPROVED** [review]
:rocket:

File: .golangci.yaml (hunk 2 above)
Comment (ti-mo) [reply .golangci.yaml:16]: Done.

File: Makefile (hunk 1 above)
Comment (joestringer) [reply Makefile]: :+1: fair enough. I'm always worried about "just a couple more minutes" on the overall build time explicitly because we mostly just tolerate the test length and we don't grapple with how to make it shorter. So any time a PR like this comes up, the easiest solution is just to tolerate making our dev cycle even longer.

That said, in general we're currently reviewing the jenkins tests for which ones we can convert to other forms of testing / delete so we should be able to make up for this additional test run time pretty soon:tm:. And yeah it's not really fair to hold up this particular PR if this is a general problem with the testing code in Cilium.

> For me, the next step would be cutting out ginkgo from running privileged tests on Jenkins. As far as I can see, it's currently only used for invoking vagrant. The code is inscrutable at best, and heavily interferes in runtime test output, especially when failures occur.

I think the only reason we do it on Jenkins today is to be able to run the tests in a privileged environment. Wherever we can get that, we can run it there.

**Review by sayboras — APPROVED** [review]

**Review by tklauser — APPROVED** [review]

**Review by gandro — APPROVED** [review]
Looks good for the changes to my codeowner groups. I did not do a review of the overall new mechanism and if it makes sense.

**Comment by ti-mo** [comment]
/test-1.23-5.4

**Comment by joestringer** [comment]
@ti-mo could you do a scan through the files changed here and double-check that the copyright notices are on the files? I found one instance `pkg/testutils/privileged.go` where the file was missing the standard SPDX/etc preamble.

**Comment by rolinh** [comment]
@ti-mo @joestringer [go-header](https://github.com/denis-tingaikin/go-header), part of the [golangci-lint](https://github.com/golangci/golangci-lint) suite can be used for this purpose. The main caveat for using in this repo is that I'm not sure we can make it work with multiple license headers.

**Comment by ti-mo** [comment]
> @ti-mo @joestringer [go-header](https://github.com/denis-tingaikin/go-header), part of the [golangci-lint](https://github.com/golangci/golangci-lint) suite can be used for this purpose. The main caveat for using in this repo is that I'm not sure we can make it work with multiple license headers.

Cool, this is relatively easy to implement ourselves, though.

**Comment by joestringer** [comment]
I thought we had a linter for this, but evidently not. It's possible that the [FOSSA scanner work](https://github.com/cilium/cilium/pull/20776) will introduce one, but it should be easy enough to add one ourselves anyway.
